smithy-annotation-processor/src/main/java/software/amazon/smithy/processor/SmithyAnnotationProcessor.java

```diff
@@ -0,0 +1,219 @@
+ * <strong>a single build plugin</strong> and write any generated Java code created
+ * by the plugin to the correct output path.
+ *
+ * <p>Smithy models on the classpath of extending annotation processors are automatically
+ * discovered and any java artifacts generated by this plugin are written to standard location.
+ * Generated resource files are written to the class output while all other generated Java
+ * classes are written to the source output. Plugin artifacts that are not java files or
+ * resource files under {@code META-INF/} are ignored.
+ *
+ * <p>Extending classes must specify the name of the build plugin they will execute using
+ * {@link #getPluginName()}. This plugin must be discoverable on the annotation processor classpath.
+ * The plugin is then executed using the data from the annotation as the plugin configuration.
+ *
+ * <p>For example, if your plugin has the following configuration node in a `smithy-build.json` file:
+ * <pre>
+ *     "myPlugin" : {
+ *         "packageName" = "myPackage.namespace",
+ *         "listOfTags" = ["a", "b", "c"]
+ *     }
+ * </pre>
+ * Then you would define an annotation like the following:
+ * <pre>
+ * {@code
+ *  @Target(ElementType.PACKAGE)
+ *  public @interface MyPlugin {
+ *     String packageName();
+ *     String[] listOfTags();
+ *  }
+ * }
+ * </pre>
+ * The {@link #createPluginNode} method is used to map this annotation to the object node
+ * expected by the plugin. Once a mapping has been defined for the annotation, it can be
+ * added to the {@code package-info.java} as follows:
+ * <pre>
+ * {@code
+ * @MyPlugin(
+ *      packageName = "myPackage.namespace",
+ *      listOfTags = {"a", "b", "c"}
+ * )
+ * package com.example.traitcodegen;
+ *
+ * import com.example.annotations.MyPlugin;
+ * }
+ * </pre>
+ *
+ * @param <A> Annotation to execute the processor for. This annotation should
+ *            target the {@code PACKAGE} {@link java.lang.annotation.ElementType} and
+ *            should be applied in the {@code package-info.java} file of a package.
+ */
+public abstract class SmithyAnnotationProcessor<A extends Annotation> extends AbstractProcessor {
+    private static final String MANIFEST_PATH = "META-INF/smithy/manifest";
+    private Messager messager;
+    private Filer filer;
+
+    @Override
+    public synchronized void init(ProcessingEnvironment processingEnv) {
+        super.init(processingEnv);
+        messager = processingEnv.getMessager();
+        filer = processingEnv.getFiler();
+    }
+
+    @Override
+    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
+        Set<? extends Element> elements = roundEnv.getElementsAnnotatedWith(getAnnotationClass());
+        if (elements.size() > 1) {
+            messager.printMessage(Diagnostic.Kind.ERROR,
+                    "Only one package can have the " + getAnnotationClass() + " annotation.");
```

Comment:
Why?

Reply:
I was thinking that we dont want to have two of the same plugins applied twice as would likely lead to duplicate/conflicting output files which would almost certainly fail because these annotation processors write to shared standard locations. 

For example we wouldn't want to allow: 
```
@MySmithyPlugin
@MySmithyPlugin
package com.example.smithy;
```
But the following might be OK, but would already be allowed by the check performed here:
```
@MySmithyPlugin
@MyOtherSmithyPlugin
package com.example.smithy;
```